src/goto-symex/symex_main.cpp

```diff
@@ -519,3 +525,115 @@ void goto_symext::symex_step(
+static void
+collect_pointer_typed_symbols(const exprt &expr, std::set<symbol_exprt> &output)
+{
+  const symbol_exprt *symbol_expr = expr_try_dynamic_cast<symbol_exprt>(expr);
+  if(symbol_expr && can_cast_type<pointer_typet>(symbol_expr->type()))
+  {
+    output.insert(*symbol_expr);
+  }
+
+  for(const exprt &op : expr.operands())
+  {
+    collect_pointer_typed_symbols(op, output);
+  }
+}
+
+void goto_symext::try_filter_value_sets(
+  goto_symex_statet &state,
+  exprt condition,
+  value_sett *value_set_for_true_case,
+  value_sett *value_set_for_false_case,
+  const namespacet &ns)
+{
+  condition = state.rename<goto_symex_statet::L1>(std::move(condition), ns);
+
+  std::set<symbol_exprt> pointer_typed_symbols{};
+  collect_pointer_typed_symbols(condition, pointer_typed_symbols);
+
+  if(pointer_typed_symbols.size() == 1)
+  {
+    const symbol_exprt &symbol_expr = *pointer_typed_symbols.begin();
+    const pointer_typet &symbol_type = to_pointer_type(symbol_expr.type());
+
+    // If symbol_expr has already been L2-renamed then we need to strip it back
+    // to L1-renaming for the purposes of getting its value set
+    const symbol_exprt symbol_expr_l1 =
+      is_ssa_expr(symbol_expr) ? to_ssa_expr(symbol_expr).get_l1_object()
+                               : symbol_expr;
+    value_setst::valuest possible_values;
+    state.value_set.get_value_set(symbol_expr_l1, possible_values, ns);
+
+    // Try evaluating the condition with symbol_expr pointing to each one of
+    // its possible values in turn. If that leads to a true from o1 then we can
+    // delete that o1 from the value set that will be used if the condition is
+    // false, and vice versa.
+    for(const auto &possible_value : possible_values)
+    {
+      if(!can_cast_expr<object_descriptor_exprt>(possible_value))
+      {
+        continue;
+      }
+
+      object_descriptor_exprt o = to_object_descriptor_expr(possible_value);
+
+      const exprt &root_object = o.root_object();
+      const exprt &object = o.object();
+      exprt value_for_symbol_expr;
+
+      if(root_object.id() == ID_null_object)
+      {
+        value_for_symbol_expr = null_pointer_exprt{symbol_type};
+      }
+      else if(root_object.id() == ID_integer_address)
+      {
+        continue;
+      }
+      else
+      {
+        // We should do something like
+        // value_set_dereference::dereference_type_compare, which deals with
+        // arrays having types containing void
+        if(o.offset().is_zero() && object.type() == symbol_type.subtype())
+        {
+          value_for_symbol_expr = address_of_exprt(object);
+        }
+        else
+        {
+          continue;
+        }
+      }
+
+      exprt modified_condition(condition);
+
+      replace_symbolt replace_symbol{};
+      replace_symbol.insert(symbol_expr, value_for_symbol_expr);
+      replace_symbol(modified_condition);
+
+      do_simplify(modified_condition);
```

Comment:
I don't think simplify-before-rename will gain us anything

Reply:
Actually I added this for a reason. When the condition is `*p = a`, `replace_symbol` makes it `*&a = a`, then `rename` makes it `a = 2` (because `rename` doesn't do constant propagation inside addresses), and then the final simplify didn't do anything and we couldn't tell if it was true or false. Perhaps there is a better way to deal with this case.

Reply:
Would you mind adding a comment to the code to this effect? It does make me wonder whether this isn't just re-implementing dereferencing (i.e., `dereference_rec`) with some extra constraints.

Reply:
That was how I first attempted to implement it. I had a class which took the place of `symex_dereference_statet`, which returned a given value for a given key and "not found" for everything else. But it was quite a pain to pass it around everywhere, and it seemed simple enough to just replace the symbol directly. Maybe it would have been better to use `dereference_rec` instead.